File: backend/src/main/java/com/taptaza/service/CompanyService.java
package com.taptaza.service;

import com.taptaza.dto.response.CompanyDetailResponse;
import com.taptaza.dto.response.CompanyResponse;
import com.taptaza.dto.response.PageResponse;
import com.taptaza.dto.response.ReviewResponse;
import com.taptaza.dto.response.ServiceResponse;
import org.springframework.data.domain.Pageable;

import java.util.List;
import java.util.UUID;

public interface CompanyService {

    /**
     * Search companies with pagination
     *
     * @param search   optional search term to filter companies by name or description
     * @param pageable pagination parameters
     * @return paginated list of companies
     */
    PageResponse<CompanyResponse> getCompanies(String search, Pageable pageable);

    /**
     * Get company details with services
     *
     * @param id company UUID
     * @return company details including services
     */
    CompanyDetailResponse getCompanyById(UUID id);

    /**
     * Get services for a company
     *
     * @param companyId company UUID
     * @return list of services offered by the company
     */
    List<ServiceResponse> getCompanyServices(UUID companyId);

    /**
     * Get reviews for a company with pagination
     *
     * @param companyId company UUID
     * @param pageable  pagination parameters
     * @return paginated list of reviews
     */
    PageResponse<ReviewResponse> getCompanyReviews(UUID companyId, Pageable pageable);
}

File: frontend/src/services/api/companyService.ts
import apiClient from './client';
import {
  CompanyResponse,
  CompanyDetailResponse,
  PageResponse,
} from './types';

class CompanyService {
  /**
   * Get list of companies with optional filters
   */
  async getCompanies(params?: {
    page?: number;
    size?: number;
    search?: string;
    minRating?: number;
    verified?: boolean;
    sortBy?: 'rating' | 'price' | 'reviews';
  }): Promise<PageResponse<CompanyResponse>> {
    const response = await apiClient.get<PageResponse<CompanyResponse>>('/companies', {
      params: {
        page: params?.page ?? 0,
        size: params?.size ?? 10,
        search: params?.search,
        minRating: params?.minRating,
        verified: params?.verified,
        sortBy: params?.sortBy,
      },
    });
    return response.data;
  }

  /**
   * Get company details by ID
   */
  async getCompanyById(id: string): Promise<CompanyDetailResponse> {
    const response = await apiClient.get<CompanyDetailResponse>(`/companies/${id}`);
    return response.data;
  }

  /**
   * Get featured/top companies
   */
  async getFeaturedCompanies(): Promise<CompanyResponse[]> {
    const response = await apiClient.get<CompanyResponse[]>('/companies/featured');
    return response.data;
  }

  /**
   * Search companies by name or service
   */
  async searchCompanies(query: string): Promise<CompanyResponse[]> {
    const response = await apiClient.get<CompanyResponse[]>('/companies/search', {
      params: { q: query },
    });
    return response.data;
  }
}

export const companyService = new CompanyService();
export default companyService;
